icicle/tests/test_hash_api.cpp

```diff
@@ -561,200 +561,206 @@ TEST_F(HashApiTest, MerkleTreeBasic)
 
 TEST_F(HashApiTest, MerkleTreeZeroPadding)
 {
-  // TODO:add loop on devices (and change hash to one supported on gpu)
-  const int leaf_size = sizeof(uint32_t);
-  const int nof_leaves = 100;
-  uint32_t leaves[nof_leaves];
-  randomize(leaves, nof_leaves);
-  ICICLE_CHECK(icicle_set_device(IcicleTestBase::reference_device()));
+  for (const auto& device : s_registered_devices) {
+    ICICLE_LOG_INFO << "MerkleTreeZeroPadding test on device=" << device;
+    ICICLE_CHECK(icicle_set_device(device));
 
-  // define the merkle tree
-  auto layer0_hash = HashSumBackend::create(5 * leaf_size, 2 * leaf_size); // in 5 leaves, out 2 leaves 400B -> 160B
-  auto layer1_hash = HashSumBackend::create(4 * leaf_size, leaf_size);     // in 4 leaves, out 1 leaf   160B ->  40B
-  auto layer2_hash = HashSumBackend::create(leaf_size, leaf_size);         // in 1 leaf, out 1 leaf     40B  ->  40B
-  auto layer3_hash = HashSumBackend::create(10 * leaf_size, leaf_size);    // in 10 leaves, out 1 leaf     40B  ->   4B
+    const int leaf_size = 256;
+    const int nof_leaves = 100;
+    std::byte leaves[nof_leaves * leaf_size];
+    randomize(leaves, nof_leaves);
 
-  int total_nof_input_hashes = nof_leaves * leaf_size / layer0_hash.default_input_chunk_size();
-  std::vector<Hash> hashes = {layer0_hash, layer1_hash, layer2_hash, layer3_hash};
-  int output_store_min_layer = 0;
+    // define the merkle tree
+    auto layer0_hash = Keccak256::create(leaf_size);
+    auto layer1_hash = Keccak256::create(2 * 32);
+    auto layer2_hash = Keccak256::create(5 * 32);
+    auto layer3_hash = Keccak256::create(10 * 32);
 
-  auto config = default_merkle_tree_config();
-  // Test zero padding
-  config.padding_policy = PaddingPolicy::ZeroPadding;
+    int total_nof_input_hashes = nof_leaves * leaf_size / layer0_hash.default_input_chunk_size();
+    std::vector<Hash> hashes = {layer0_hash, layer1_hash, layer2_hash, layer3_hash};
+    int output_store_min_layer = 0;
 
-  ICICLE_LOG_DEBUG << "Full tree";
-  test_merkle_tree(hashes, config, output_store_min_layer, nof_leaves, leaves);
+    auto config = default_merkle_tree_config();
+    // Test zero padding
+    config.padding_policy = PaddingPolicy::ZeroPadding;
+
+    ICICLE_LOG_VERBOSE << "Full tree";
+    test_merkle_tree(hashes, config, output_store_min_layer, nof_leaves, leaves);
+
+    ICICLE_LOG_VERBOSE << "Last hash isn't full";
+    test_merkle_tree(hashes, config, output_store_min_layer, nof_leaves - 1, leaves);
+
+    const unsigned nof_leaves_in_hash = layer0_hash.default_input_chunk_size() / leaf_size;
+
+    ICICLE_LOG_VERBOSE << "19 hashes (Total hashes in layer 0 - 1) - full";
```

Comment:
Can you remove the 19 here? the changes made in the tree above now yield 100 hashes in the lowest level instead of the previous 20.

Reply:
I refined the tests. Seems like there is a lot of redundancy